l1-contracts/contracts/bridge/ntv/NativeTokenVault.sol

```diff
@@ -0,0 +1,419 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.24;
+
+// solhint-disable reason-string, gas-custom-errors
+
+import {Ownable2StepUpgradeable} from "@openzeppelin/contracts-upgradeable-v4/access/Ownable2StepUpgradeable.sol";
+import {PausableUpgradeable} from "@openzeppelin/contracts-upgradeable-v4/security/PausableUpgradeable.sol";
+import {BeaconProxy} from "@openzeppelin/contracts-v4/proxy/beacon/BeaconProxy.sol";
+import {IBeacon} from "@openzeppelin/contracts-v4/proxy/beacon/IBeacon.sol";
+
+import {IERC20} from "@openzeppelin/contracts-v4/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts-v4/token/ERC20/utils/SafeERC20.sol";
+
+import {IBridgedStandardToken} from "../interfaces/IBridgedStandardToken.sol";
+import {INativeTokenVault} from "./INativeTokenVault.sol";
+import {IAssetHandler} from "../interfaces/IAssetHandler.sol";
+import {IAssetRouterBase} from "../asset-router/IAssetRouterBase.sol";
+import {DataEncoding} from "../../common/libraries/DataEncoding.sol";
+import {ETH_TOKEN_ADDRESS} from "../../common/Config.sol";
+
+import {BridgedStandardERC20} from "../BridgedStandardERC20.sol";
+import {BridgeHelper} from "../BridgeHelper.sol";
+
+import {EmptyDeposit, Unauthorized, TokensWithFeesNotSupported, TokenNotSupported, NonEmptyMsgValue, ValueMismatch, WithdrawFailed, InsufficientChainBalance, AddressMismatch, AssetIdMismatch, AmountMustBeGreaterThanZero} from "../../common/L1ContractErrors.sol";
+
+/// @author Matter Labs
+/// @custom:security-contact security@matterlabs.dev
+/// @dev Vault holding L1 native ETH and ERC20 tokens bridged into the ZK chains.
+/// @dev Designed for use with a proxy for upgradability.
+abstract contract NativeTokenVault is INativeTokenVault, IAssetHandler, Ownable2StepUpgradeable, PausableUpgradeable {
+    using SafeERC20 for IERC20;
+
+    /// @dev Contract that stores the implementation address for token.
+    /// @dev For more details see https://docs.openzeppelin.com/contracts/3.x/api/proxy#UpgradeableBeacon.
+    IBeacon public bridgedTokenBeacon;
+
+    /// @dev The address of the WETH token.
+    address public immutable override WETH_TOKEN;
+
+    /// @dev L1 Shared Bridge smart contract that handles communication with its counterparts on L2s
+    IAssetRouterBase public immutable override ASSET_ROUTER;
+
+    /// @dev Maps token balances for each chain to prevent unauthorized spending across ZK chains.
+    /// This serves as a security measure until hyperbridging is implemented.
+    /// NOTE: this function may be removed in the future, don't rely on it!
+    mapping(uint256 chainId => mapping(address l1Token => uint256 balance)) public chainBalance;
+
+    /// @dev A mapping assetId => tokenAddress
+    mapping(bytes32 assetId => address tokenAddress) public tokenAddress;
+
+    /// @dev A mapping assetId => isTokenBridged
+    mapping(bytes32 assetId => bool bridged) public isTokenBridged; // kl todo should we have isTokenNativeInstead
```

Comment:
also, do we need to have a way to migrate all the old tokens into the `isTokenBridged` mapping?